File: leetcode-solution/leetcode114/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */


public class Solution {
    //定义一个lastNode来装在flatten后的数据
    private TreeNode lastNode = null;
    
    /*
     * @param root: a TreeNode, the root of the binary tree
     * @return: 
     */
    public void flatten(TreeNode root) {
        //write your code here
        //如果我碰到一个节点，那么我就把它的父节点的左孩子改为null，右儿子改
        //为当前节点，递归遍历下去。lastnode指向的是上一个处理的节点，也可以
        //看成是当前要处理的节点在“假链表”中的父节点。
        //flatten左子树，把左子树最下最右的一点，接到右子树的root
        //上，然后把root接到左子树的root前即可
        if (root == null) {
            return; 
        }
        //如果lastNode不是空的，那么把其左子树删光，把其右子树接上原树的root
        if (lastNode != null) {
            lastNode.left = null;
            lastNode.right = root;
        }
        //如果lastNode是空的，那么把其接上原树的root
        lastNode = root;
        //就是在flatten(root.left);, 过程中root.right 发生了改变，在这行
        //lastNode.right = root;，所以我们需要先记录一下right作为临时变量。
        //可以画图看看
        TreeNode right = root.right;
        //再对原树的左子树做同样的事
        flatten(root.left);
        flatten(right); 
    }
}


public class Solution {
     /*
     * @param root: a TreeNode, the root of the binary tree
     * @return: 
     */
    public void flatten(TreeNode root) {
        //首先要理解为什么要Last？因为我们要合并两部分内容，怎么合并的，就是
        //左边的last 后边接上右边的第一个. 所以我们返回的last是用于知道当前
        //展开后的最后一个元素，既然右边的是接在左边的后面，那么右边的存在，
        //不就是最后的一定来自于右边么，如果右边的不存在，那么他来自于左边的
        //最后一个。
        helper(root);
    }
    private TreeNode helper(TreeNode root) {
        if (root == null) {
            return null;
        }
        //代表左右边的最后一个,分治
        TreeNode leftLast = helper(root.left);
        TreeNode rightLast = helper(root.right);
        //只要左边的最后一个不是空的，然后将leftLast接上原树root右边的最开
        //始
        if (leftLast != null) {
            leftLast.right = root.right;
            //root的right一边连上root的left，就是把root的left换到root的right
            //的位置
            root.right = root.left;
            //使左边成为null
            root.left = null;
        }
        //这两个if语句不能换，你可能没有理解这个函数在做什么.首先这个函数返回
        //的是flatten之后链表的最后一个元素，那么如果rightLast 存在，那么当
        //前的最后一个元素就是rightLast，只有不存在的时候，leftLast才可能是
        //最后的元素，也就是前面存在了，后面if语句是否成立无关紧要。后面if语
        //句的存在，依赖于前面的if语句。
        if (rightLast != null) {
            return rightLast;
        }
        if (leftLast != null) {
            return leftLast;
        }
        return root;
    }
}

File: leetcode-solution/leetcode114/solution.js
// Flatten Binary Tree to Linked List
// Given a binary tree, flatten it to a linked list in-place.

// For example, given the following tree:

//     1
//    / \
//   2   5
//  / \   \
// 3   4   6

// The flattened tree should look like:

// 1
//  \
//   2
//    \
//     3
//      \
//       4
//        \
//         5
//          \
//           6

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * param {TreeNode} root
 * return {void} Do not return anything, modify root in-place instead.
 */
// 思路：
// 1. 用前序遍历遍历出原tree
// 2. root.left将root的左子树设为null
// 3. cursor这时候得到了左子树为空的root
// 4. for循环list里的用前序遍历得到的root
// 5. cursor的右子节点设为每次list里的一项（这里如果不用treenode，无法放到cursor里，因为cursor里的是一个树
// 6. 然后将cursor的根节点设为其右子节点，这样循环，就能把list里的每一项一次一次地接在右子节点后面
var flatten = function(root) {
    if (!root) {
        return [];
    }
    let list = [];
    DFS(root, list);
    root.left = null;
    let cursor = root;
    for (let i = 1; i <= list.length - 1; i++) {
        cursor.right = new TreeNode(list[i].val);
        cursor = cursor.right;
    }
};

//preorder traversal;
function DFS (root, list) {
    if (!root) {
        return [];
    }
    list.push(root);
    if (root.left) {
        DFS(root.left, list);
    }
    if (root.right) {
        DFS(root.right, list);
    }
    return list;
}